File: verifier-go/xcs/lifecycle.go
package xcs

// CredentialLifecycleState is the state projected from a validated ledger object.
type CredentialLifecycleState string

const (
	CredentialLifecyclePending CredentialLifecycleState = "pending"
	CredentialLifecycleActive  CredentialLifecycleState = "active"
	CredentialLifecycleExpired CredentialLifecycleState = "expired"
	CredentialLifecycleDeleted CredentialLifecycleState = "deleted"
)

// CredentialLifecycleInput contains only authoritative ledger lifecycle evidence.
type CredentialLifecycleInput struct {
	ObjectExists bool
	Accepted     bool
	Expiration   *uint32
	CloseTime    uint32
}

// ProjectCredentialLifecycle applies the protocol's lifecycle precedence rules.
func ProjectCredentialLifecycle(input CredentialLifecycleInput) CredentialLifecycleState {
	if !input.ObjectExists {
		return CredentialLifecycleDeleted
	}
	if input.Expiration != nil && *input.Expiration <= input.CloseTime {
		return CredentialLifecycleExpired
	}
	if input.Accepted {
		return CredentialLifecycleActive
	}
	return CredentialLifecyclePending
}

File: packages/core/src/lifecycle.ts
import { fail } from './errors.js'
import type { CredentialLifecycleInput, CredentialLifecycleState } from './types.js'

const UINT32_MAX = 0xffff_ffff

function requireRippleTime(value: number): void {
  if (!Number.isInteger(value) || value < 0 || value > UINT32_MAX) {
    return fail('RIPPLE_TIME_INVALID', 'Credential lifecycle time must be a uint32', '$time')
  }
}

/** Projects the state of a known Credential generation at a validated-ledger close time. */
export function projectCredentialLifecycle(
  input: CredentialLifecycleInput,
): CredentialLifecycleState {
  requireRippleTime(input.closeTime)
  if (input.expiration !== undefined && input.expiration !== null) {
    requireRippleTime(input.expiration)
  }

  if (!input.objectExists) return 'deleted'
  if (
    input.expiration !== undefined &&
    input.expiration !== null &&
    input.expiration <= input.closeTime
  )
    return 'expired'
  return input.accepted ? 'active' : 'pending'
}
